File: 프로그래머스/2/87946. 피로도/피로도.py
def solution(k, dungeons):
    max_count = 0
    visited = [False] * len(dungeons)
    
    def find_dungeons (count, nowK, visited):
        nonlocal max_count
        max_count = max(count, max_count)
        
        for i in range(len(dungeons)):
            if not visited[i] and dungeons[i][0] <= nowK:
                next_nowK = nowK - dungeons[i][1]
                visited[i] = True
                find_dungeons (count + 1, next_nowK, visited)
                visited[i] = False
        
    find_dungeons(0, k, visited)
    
    return max_count

File: 프로그래머스/2/87946. 피로도/피로도.js
function solution(k, dungeons) {
    const visited = Array(dungeons.length).fill(false);
    let count = 0;
    let nowK = k;
    
    let maxCount = 0;
    function findDungeons(count, nowK, visited) {        
        
        maxCount = Math.max(count, maxCount);
        
        for (let i = 0; i < dungeons.length; i++) {
            if (!visited[i] && nowK >= dungeons[i][0]) {
                let nextNowK = nowK -  dungeons[i][1];
                visited[i] = true;
                findDungeons(count + 1, nextNowK, visited);
                visited[i] = false;
            }
        }
    }
    
    findDungeons(count, k, visited);
    
    return maxCount;
}